src/supernova/mod.rs

```diff
@@ -813,13 +815,74 @@ where
   (ck_primary, ck_secondary)
 }
 
-/// SuperNova helper trait, for implementors that provide sets of sub-circuits to be proved via NIVC.
-pub trait CircuitSet<G: Group> {
+/// SuperNova helper trait, for implementors that provide sets of sub-circuits to be proved via NIVC. `C1` must be a
+/// type (likely an `Enum`) for which a potentially-distinct instance can be supplied for each `index` below
+/// `self.num_circuits()`.
+pub trait NonUniformCircuit<G1, G2, C1>
+where
+  G1: Group<Base = <G2 as Group>::Scalar>,
+  G2: Group<Base = <G1 as Group>::Scalar>,
+  C1: StepCircuit<G1::Scalar>,
+{
   /// Initial program counter, defaults to zero.
-  fn initial_program_counter(&self) -> G::Scalar {
-    G::Scalar::ZERO
+  fn initial_program_counter(&self) -> G1::Scalar {
+    G1::Scalar::ZERO
+  }
+
+  /// Return the initial running claims for `NonUniformCircuit`'s sub-circuits.
+  fn initial_running_claims(
+    &self,
+  ) -> Vec<RunningClaim<G1, G2, C1, TrivialSecondaryCircuit<G2::Scalar>>> {
```

Comment:
By 1.74, we should be able to make this return an `impl IntoIterator<Item = ...>`, instead of this form, which forces an allocation of an intermediate result that's never used on its own. The feature is `return_position_impl_trait_in_trait`